### ISSUE TITLE
error when training, selected index k out of range

### ISSUE THREAD

**Issue by cena001plus** [Megvii-BaseDetection/YOLOX#778]
this issue is same to #277， but i have not get useful answer, 80 epoch, when it runs to the 65th epoch, which is the 15th epoch from the bottom, an error occurs at this time. I read the related issue #277, saying that it is to modify the n_candidate_k = min(10, ious_in_boxes_matrix.size) in line 619. (1)), but what I downloaded is the latest version. The name is already in line 613. What is the situation? and my class number is 3, why gt_classes=9? The part log is as follows:

``------------CPU Mode for This Batch-------------
2021-10-09 13:13:19 | ERROR    | yolox.models.yolo_head:328 - OOM RuntimeError is raised due to the huge memory cost during label assignment.                            CPU mode is applied in this batch. If you want to avoid this issue,                            try to reduce the batch size or image size.
------------CPU Mode for This Batch-------------
------------CPU Mode for This Batch-------------
2021-10-09 13:13:19 | INFO     | yolox.core.trainer:184 - Training of experiment is done and the best AP is 1.85
2021-10-09 13:13:19 | ERROR    | yolox.core.launch:147 - An error has been caught in function '_distributed_worker', process 'SpawnProcess-1' (1487502), thread 'MainThread' (140675952382976):
Traceback (most recent call last):

File "/media/D/wuxl/work/my_prj/YOLOX/yolox/models/yolo_head.py", line 505, in get_assignments
    ) = self.dynamic_k_matching(cost, pair_wise_ious, gt_classes, num_gt, fg_mask)
        │    │                  │     │               │           │       └ tensor([False, False, False,  ..., False, False, False], device='cuda:0')
        │    │                  │     │               │           └ 9
        │    │                  │     │               └ tensor([2., 2., 2., 2., 0., 0., 0., 0., 1.], device='cuda:0')
        │    │                  │     └ tensor([], device='cuda:0', size=(9, 0))
        │    │                  └ tensor([], device='cuda:0', size=(9, 0))
        │    └ <function YOLOXHead.dynamic_k_matching at 0x7ff110082680>
        └ YOLOXHead(
            (cls_convs): ModuleList(
              (0): Sequential(
                (0): BaseConv(
                  (conv): Conv2d(128, 128, kernel_siz

**Comment by Darknesszlx** [comment]
https://github.com/Megvii-BaseDetection/YOLOX/blob/12d0df18b5f5010ed264c499d069fc107f6e85cc/yolox/models/yolo_head.py#L576
I fix the bug by enlarging center_radius value, you can try 3 , 3.5... or other values bigger than 2.5

**Comment by rsomani95** [comment]
@cena001plus @Darknesszlx were you guys training on a custom dataset?
I'm trying to train a COCO model without padding the images (#1063), and am running into this

Using @ahmedanis03's snippet above, I can see that there's many iterations where no matches are found, _despite_ changing the `center_radius` to `[3, 3.5, 4, 5]`, and eventually I run out of memory during label assignment and the training run crashes

**Comment by Darknesszlx** [comment]
> @cena001plus @Darknesszlx were you guys training on a custom dataset? I'm trying to train a COCO model without padding the images (#1063), and am running into this
> 
> Using @ahmedanis03's snippet above, I can see that there's many iterations where no matches are found, _despite_ changing the `center_radius` to `[3, 3.5, 4, 5]`, and eventually I run out of memory during label assignment and the training run crashes

I trained on my own dataset. The problem is caused by the situation that predicted boxes can't be assigned to any gt boxes when gt box num is bigger than 0 (positive training sample), so you can filter out those hard positive training samples, or find another way to make sure label assignment runs OK(my dataset works well when I enlarged center_radius).

**Comment by fanfanfan111** [comment]
> > 更改以下内容
> > ```
> > def dynamic_k_matching(self, cost, pair_wise_ious, gt_classes, num_gt, fg_mask):
> >         # Dynamic K
> >         # ---------------------------------------------------------------
> >         matching_matrix = torch.zeros_like(cost)
> > 
> >         ious_in_boxes_matrix = pair_wise_ious
> >         n_candidate_k = min(10, ious_in_boxes_matrix.size(1))
> >         if n_candidate_k != 0:
> >             #print(pair_wise_ious, gt_classes, num_gt)
> >             topk_ious, _ = torch.topk(ious_in_boxes_matrix, n_candidate_k, dim=1)
> >             dynamic_ks = torch.clamp(topk_ious.sum(1).int(), min=1)
> >             for gt_idx in range(num_gt):
> >                 _, pos_idx = torch.topk(
> >                     cost[gt_idx], k=dynamic_ks[gt_idx].item(), largest=False
> >                 )
> >                 matching_matrix[gt_idx][pos_idx] = 1.0
> > 
> >             del topk_ious, dynamic_ks, pos_idx
> >         else:
> >             print("Sth wrong happened!")
> >        ...rest of the function as is outside the if condition
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > 这是一个虚拟的解决方案。根本原因是模型没有预测给定观察中的任何框，这似乎是不合逻辑的行为，因为它不是由开发人员处理的，这也表明模型没有学到任何东西，正如您在案例中看到的那样AP是1.85，我的也是。
> > 所以这将使它向前发展，但最终你仍然会得到一个什么都没学到的虚拟模型。
> > 顺便说一句，我在同一个问题上尝试了所有相同的东西（模型配置、gpu、fp16）等，但在不同的数据集上，它工作正常，所以很可能这是一个数据集问题。
> > 我将进行更多调查并更新问题。
> 
> 你是对的，只是换了voc2007数据集，整个训练结果是正确的，换了我们自己的数据集，就出现了这个问题，训练效果极差，但是同样的数据集在YOLOv4-tiny上工作正常。所以我觉得数据集比较特殊，但主要问题应该是代码。#565

你好 在自己建立的数据集上 最后15轮关闭马赛克增强的时候出现了一样的错误，请问是数据集的关系吗？

**Comment by LordonCN** [comment]
> > > 更改以下内容
> > > ```
> > > def dynamic_k_matching(self, cost, pair_wise_ious, gt_classes, num_gt, fg_mask):
> > >         # Dynamic K
> > >         # ---------------------------------------------------------------
> > >         matching_matrix = torch.zeros_like(cost)
> > > 
> > >         ious_in_boxes_matrix = pair_wise_ious
> > >         n_candidate_k = min(10, ious_in_boxes_matrix.size(1))
> > >         if n_candidate_k != 0:
> > >             #print(pair_wise_ious, gt_classes, num_gt)
> > >             topk_ious, _ = torch.topk(ious_in_boxes_matrix, n_candidate_k, dim=1)
> > >             dynamic_ks = torch.clamp(topk_ious.sum(1).int(), min=1)
> > >             for gt_idx in range(num_gt):
> > >                 _, pos_idx = torch.topk(
> > >                     cost[gt_idx], k=dynamic_ks[gt_idx].item(), largest=False
> > >                 )
> > >                 matching_matrix[gt_idx][pos_idx] = 1.0
> > > 
> > >             del topk_ious, dynamic_ks, pos_idx
> > >         else:
> > >             print("Sth wrong happened!")
> > >        ...rest of the function as is outside the if condition
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > 这是一个虚拟的解决方案。根本原因是模型没有预测给定观察中的任何框，这似乎是不合逻辑的行为，因为它不是由开发人员处理的，这也表明模型没有学到任何东西，正如您在案例中看到的那样AP是1.85，我的也是。
> > > 所以这将使它向前发展，但最终你仍然会得到一个什么都没学到的虚拟模型。
> > > 顺便说一句，我在同一个问题上尝试了所有相同的东西（模型配置、gpu、fp16）等，但在不同的数据集上，它工作正常，所以很可能这是一个数据集问题。
> > > 我将进行更多调查并更新问题。
> > 
> > 
> > 你是对的，只是换了voc2007数据集，整个训练结果是正确的，换了我们自己的数据集，就出现了这个问题，训练效果极差，但是同样的数据集在YOLOv4-tiny上工作正常。所以我觉得数据集比较特殊，但主要问题应该是代码。#565
> 
> 你好 在自己建立的数据集上 最后15轮关闭马赛克增强的时候出现了一样的错误，请问是数据集的关系吗？

Yes, I also have this problem and I find the value of pair_wise_ious.size(1) is Zero.
